typing/typecore.ml

```diff
@@ -3148,42 +3147,46 @@ and type_expect_
             type_let existential_context env rec_flag spat_sexp_list
               allow_modules
           in
-          let body = type_expect new_env sbody ty_expected_explained in
-          let () =
-            if rec_flag = Recursive then
-              check_recursive_bindings env pat_exp_list
-          in
-          (* The "bound expressions" component of the scope escape check.
+          if rec_flag = Recursive then
+            check_recursive_bindings env pat_exp_list;
 
+          let body = type_expect new_env sbody ty_expected_explained
```

Comment:
The test `letrec-check/pr7231.ml` has changed output. The output change seems acceptable &mdash; previously, a warning was emitted before a fatal error, and now we just get the fatal error &mdash; but we can easily get the old output by re-swapping the calls to `type_expect` and `check_recursive_bindings`.

```
--- /home/runner/work/ocaml/ocaml/testsuite/tests/letrec-check/pr7231.ocaml.reference	2023-06-26 05:02:02.000000000 +0000
+++ /home/runner/work/ocaml/ocaml/testsuite/tests/letrec-check/_ocamltest/tests/letrec-check/pr7231/ocaml/ocaml.output	2023-06-26 05:15:18.835490961 +0000
@@ -1,8 +1,3 @@
-Line 5, characters 58-64:
-5 | let rec r = let rec x () = r and y () = x () in y () in r "oops";;
-                                                              ^^^^^^
-Warning 20 [ignored-extra-argument]: this argument will not be used by the function.
-
 Line 5, characters 12-52:
 5 | let rec r = let rec x () = r and y () = x () in y () in r "oops";;
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

My vote would be for swapping the lines to maintain the old output, but I don't have objections to updating the test output.

Reply:
Personally I would be absolutely fine with the code and output change if there was some discernable logic to the new ordering that made it somehow preferable to the previous one. I thought at first that the idea was to first write all the code about `pat_exp_list` before we move to type-check `body`, and this makes a lot of sense. But this is not what the code does, as the `bound_exp_types` computation remains after `body` checking. And I cannot figure out why the `let body = ... and bound_exp_types = ...` form is used -- it makes evaluation order surprising.